### PR TITLE
Remove EventSourcedEntity, #27724

### PR DESCRIPTION
**Pull request by patriknw**
See issue #27724 for motivation.

Just a draft experiment so far because I want input on the aspect of exposing PersistenceId, comments inline.

Refs #27724

### CODE REVIEW

File: akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java
Comment (patriknw) [inline akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java]: This would be ok for Akka, but not for Play since the purpose was to hide the EntityTypeKey, so move on to the Account example instead...

File: akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithEventHandlersInState.scala
Comment (patriknw) [inline akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithEventHandlersInState.scala]: EventSourcedBehavior must have the PersistenceId so the question is how to construct it. Here is a suggestion that the EntityContext provides the "standard" PersistenceId, but it must still be passed in as a parameter by the user to the EventSourcedBehavior. Would that be ok?

I was pondering if we could make the persistenceId optional so that it could be added after construction (by the EntityFactory in Play), but that doesn't work because it should also be possible to wrap the EventSourcedBehavior in other behaviors, such as `Behaviors.withTimers` and `Behaviors.supervise`, and then the EventSourcedBehavior type is lost (it's only a Behavior).

**Review by patriknw — COMMENTED** [review]
@renatocaval I see no problem from Akka's perspective to remove EventSourcedEntity, only benefits. However, before I remove it I'd like your thoughts on how the PersistenceId should be given to the EventSourcedBehavior, see comments inline...

cc @TimMoore

**Review by octonato — COMMENTED** [review]
I'm not completely convinced by the `PersistenceId` on the `EntityContext`. It will simplify Play's factory, but there are some hidden traps in there for Akka users. 

See my comment below...

**Comment by renatocaval** [comment]
And we if only add the `EntityTypedKey` to the `EntityContext`? 

Then a user can choose which id strategy they want, for instance:

final class EntityContext(val entityId: String, typeKey: EntityTypeKey, val shard: ActorRef[ClusterSharding.ShardCommand]) {
  def entityIdTyped: String = ??? // user TypeKey to add 'type' hint on id
}

```scala
// to use when we want the original id, eg: 'abc'
EventSourcedBehavior.withEnforcedReplies(
	Persistence(entityContext.entityId), 
	EmptyAccount, 
	commandHandler, 
	eventHandler
)
```


```scala
// to use when we want a tweaked id, eg: ' $entity-name|abc'
EventSourcedBehavior.withEnforcedReplies(
	Persistence(entityContext.entityIdTyped), 
	EmptyAccount, 
	commandHandler, 
	eventHandler
)
```

We can then remove `persistenceIdFrom` method and have `entityIdTypedFrom(id: String): String`. `EntityTypedKey` will have no knowledge about `PersistenceId`, but it will help you generate an id that has the type suffix in it.

File: akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithEventHandlersInState.scala
Comment (patriknw) [reply akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithEventHandlersInState.scala]: > use PersistenceId(ctx.entityId)

yup, that is confusing. We should make it clear that the entityId is the business domain identity, as is. PersistenceId has the additional requirement to be globally unique, hence prefixing with the type name.

It's also a typical requirement that the entity itself has to know it's entityId, and therefore it can pass it from the EntityContext to the constructor.

> What I dislike in this approach is that if I'm not using an EventSourcedBehavior, I will still see a persistenceId method on my EntityContext. People will ask themselves why this exists? The same goes for EntityTypeKey.persistenceIdFrom though, so the problem is already there.

That is true, but I think we need to be somewhat pragmatic about that since we want to support the combination of ClusterSharding and EventSourcedBehavior as seamless as possible. That is the most common way to use ClusterSharding. We have to document that persistenceId can be ignored if Akka Persistence isn't used. entityId is always valid.

> If I'm a previous Akka untyped user, I will need to call it with PersistenceId(entityContext.entityId). But then I will see two methods on my EntityContext (entityId and persistenceId), and I will ask myself, why can't I simply use the persistenceId one?

I don't understand this part. `PersistenceId(entityContext.entityId)` would always be wrong.
If you have a previous persistenceId format it might be compatible with the EntityTypeKey format, with custom separator. Otherwise you have to construct it yourself from a plain `PersistenceId`, and can't use the EntityContext.persistenceId.

Maybe it should be named `EntityContext.defaultPersistenceId`.

**Comment by patriknw** [comment]
> We can then remove persistenceIdFrom method and have entityIdTypedFrom(id: String): String. EntityTypedKey will have no knowledge about PersistenceId, but it will help you generate an id that has the type suffix in it.

Yes, that could be a way to "hide" the relation to persistence, but then it becomes stringly typed instead of making use of that `PersistenceId` is a separate type.

**Comment by renatocaval** [comment]
> Yes, that could be a way to "hide" the relation to persistence, but then it becomes stringly typed instead of making use of that PersistenceId is a separate type.

But the `EventSourcedBehavior` will be requiring a `PersistenceId`. So still typed.

We will only be using the `EntityTypeKey` to generate a new String with a type hint.

File: akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithEventHandlersInState.scala
Comment (renatocaval) [reply akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithEventHandlersInState.scala]: > I don't understand this part. `PersistenceId(entityContext.entityId)` would always be wrong.

It won't be wrong with I have been using Akka Persistence (untyped) and my journal is filled with rows using the business domain id. 

My concern is that we can confuse users with a method `persistenceId`. Such a user should not use that method, but `PersistenceId(entityContext.entityId)`. 

I don't think that `defaultPersistenceId` is conveying enough meaning. That's why I was going on the direction of a method that explicitly says: "here you get you business domain id with a type hint to make it globally unique"

**Comment by renatocaval** [comment]
Hmm, I get know your argument for returning `PersistenceId` type instead of a string with a type hint. You want to avoid that this get used in all kind of behaviors. In other worlds, only an `EventSourcedBehavior` should use IDs with a type hint. 

Is that your thought/reason?

File: akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithEventHandlersInState.scala
Comment (patriknw) [reply akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithEventHandlersInState.scala]: > It won't be wrong if I have been using Akka Persistence (untyped) and my journal is filled with rows using the business domain id.

sure, but then you have selected the persistenceId badly. Orange with "123" must not have the same persistenceId as Apple "123".

**Comment by patriknw** [comment]
> But the EventSourcedBehavior will be requiring a PersistenceId. So still typed.

>  I get know your argument for returning PersistenceId type instead of a string with a type hint. You want to avoid that this get used in all kind of behaviors.

One reason was to not hide that the purpose is for EventSourcedBehavior (persistence). If you are looking for a way to construct a PersistenceId and you see a method that returns a PersistenceId that is easier to discover that those can be glued together.

**Comment by renatocaval** [comment]
@patriknw, I'm fine with those changes and it will work well with play-cqrs. It will remove the responsibility from the Factory. And it will allow users to step out of the Factory, if they want to, without any extras.

If you are ok with the risk of a previous akka persistence user to unadvisedly use  `EntityTypeKey.persistenceId`, while they need to use `PersistenceId(entityCtx.entityId)`, then I'm fine with that.

**Review by octonato — APPROVED** [review]

**Comment by patriknw** [comment]
Thanks, then I'll move forward with this.

**Comment by patriknw** [comment]
I have completed this, ready for wider review.

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala
Comment (patriknw) [inline akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala]: @renatocaval is this name ok? I was thinking that it's good to start with `persisteceId` but still give a hint about that it's optional

Comment (renatocaval) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala]: I'm fine with just `persistenceId`. I don't think 'proposal' is making it clearer. 

I find it strange to name it 'proposal' it gives the impression that's something that can be rejected. Or something that you have to branch (if/else) to make a decision.

Alternatively, we could give the responsibility to the `PersistenceId` class instead of using the `EntityTypeKey`. Correct me if I'm wrong, but there is no hard requirement that the name of the entity needs to be the same as the type hint of the persistence id. Then we have the recommended way of doing it, ie: `PersistenceId(typeHint = "Account", id = ctx.entityId)` and the legacy way `PersistenceId(ctx.entityId)`. 

And if I'm not using sharding, I get also exposed, via API, to the recommended way.

I would expect a user to define the entity key as `EntityTypeKey[AccountCommand]("Account")`, but it should also be fine if defined as `EntityTypeKey[AccountCommand]("account-model")`

Maybe we are just creating that problem because we are trying to reuse the name.

Comment (renatocaval) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala]: I just read your email where you point a previous decision to make the releation with sharding and persistence more clear. I'm aware that this is the intent and I don't want to re-open this discussion. So, please ignore my proposal to move it to the `PersistenceId`. 

Unless you find it really cool. ;-)

Comment (TimMoore) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala]: Personally, I like the suggestion from @renatocaval to have a type parameter in `PersistenceId.apply`. It's true that not only cluster sharding use cases need to disambiguate the ID.

Comment (patriknw) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala]: If we want to get rid of the persistence stuff in sharding api I think your suggestion makes sense. We would document it as the recommended way to construct a PersistenceId when used together with Sharding would be:
```
PersistenceId(typeHint = Account.TypeKey.name, id = entityContext.entityId)
```

but one can use whatever other format

I have been thinking about if the EntityTypeKey should be included in the EntityContext. Wouldn't hurt I guess, even though it's typically already in scope from where `init` is callled.

Then it could look like this:
```
PersistenceId(typeHint = entityContext.entityTypeKey.name, id = entityContext.entityId)
```




we would also have to have an optional `separator` parameter.

Comment (ignasi35) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala]: > I have been thinking about if the EntityTypeKey should be included in the EntityContext. Wouldn't hurt I guess, even though it's typically already in scope from where `init` is callled.

IMHO, it makes sense to have the `EntityTypeKey` in the `EntityContext` so the function `EntityContext => Behavior[M]` has all the information it needs. Of course users can close over the `EntityTypeKey` when creating the `EntityContext => Behavior[M]` but I found it lacking that `EntityContext` didn't give me all the _context_ual information.

> Then it could look like this:
> 
> ```
> typeHint = entityContext.entityTypeKey.name
> id = entityContext.entityId
> ```

👆🏼 This looks great.

Comment (ignasi35) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala]: (let me put it on a separate comment to keep things separate)

> `PersistenceId(typeHint = ..., id = entityContext.entityId)`

I also like that, putting sharding aside, `PersistenceId` now has an extra field.

Comment (ignasi35) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala]: Summing my two comments above: I think having access on the `EntityTypeKey` from the `EntityContext` and adding the `typeHint` on the `PersistenceId` provide a flexible enough API for users.

When we're sharding projection workers in Lagom, the `EntityTypeKey.name` has no relation with the `PersistenceId.typeHint`. Instead, `EntityTypeKey.name` is the prefix used by the tagger.

So adding this flexibility in `EntityTypeKey` and `PersistenceId` will allow a very simple integration with projections code.

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala
Comment (ignasi35) [inline akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: (related to https://github.com/akka/akka/pull/27725#discussion_r325515363)

I assume the purpose is to bring Persitence API and Sharding API closer. I wonder if these helpers `EntityContext => PersistenceId` should be part of Sharding or, instead, they should be part of Persistence. I have the gut feeling that it's Persistence that needs Sharding and not the other way around.

Similarly, Projections need Sharding and I think it'd be a mistake if we polluted `EntityContext` with `EntityContext => WorkerCoordinates` (ref [`WorkerCoordinates`](https://github.com/lagom/lagom/blob/master/projection/core/src/main/scala/com/lightbend/lagom/internal/projection/ProjectionRegistryActor.scala#L47) ).

**Comment by patriknw** [comment]
@renatocaval @ignasi35 I have completed the addition change to move things to PersistenceId.

**Review by octonato — COMMENTED** [review]
Looking good. I'm happy we moved on that direction. 

I left a small comment, just an idea. I understand that the order of parameters is mimicking the final formal, ie: hint + separator + entityId, but want to explore another ordering if it make it more natural for users.

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/PersistenceId.scala
Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/PersistenceId.scala]: Yes, let's do that since that is the normal convention of optional parameters.
My thinking was to place them in the same order as the format with the separator in the middle.

File: akka-cluster-sharding-typed/src/test/java/akka/cluster/sharding/typed/javadsl/ClusterShardingPersistenceTest.java
Comment (chbatey) [inline akka-cluster-sharding-typed/src/test/java/akka/cluster/sharding/typed/javadsl/ClusterShardingPersistenceTest.java]: could we use persistenceId().id() instead?

**Review by chbatey — APPROVED** [review]
LGTM with one question about whether we should pass around entityId in examples

File: akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java
Comment (patriknw) [reply akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java]: no, we should make it clear that they are (slightly) different. entityId is the business domain id and persistenceId is more unique journal identifier. This is what I wrote in docs:
```
The `entityId` in Cluster Sharding is the business domain identifier of the entity. The `entityId` might not
be unique enough to be used as the `PersistenceId` by itself. For example two different types of
entities may have the same `entityId`. To create a unique `PersistenceId` the `entityId` is can be prefixed
with a stable name of the entity type, which typically is the same as the `EntityTypeKey.name` that
is used in Cluster Sharding. There are @scala[`PersistenceId.apply`]@java[`PersistenceId.of`] factory methods
to help with constructing such `PersistenceId` from a `entityTypeHint` and `entityId`.
```

**Comment by patriknw** [comment]
fixed conflicts, will merge after validation

**Review by octonato — COMMENTED** [review]
I just wanted to drop another idea here before this get merged. 

All three parameters are of type String. Maybe we should have the following to make sure people don't swap them:

```scaal
PersistenceId(id)
	.withTypeHint(hint)
	.withSeparator(sep)
```

Just an idea...

**Comment by patriknw** [comment]
I think it's overkill to use a builder for those 3 params.

**Comment by renatocaval** [comment]
I was not suggesting a real builder, just some methods to have some guidance. But fair enough, the way it is now if good enough.

**Pull request merged**
